include/deal.II/lac/trilinos_tpetra_solver_direct.templates.h

```diff
@@ -0,0 +1,225 @@
+      solver->setX(x.trilinos_rcp());
+
+      // Assign the RHS vector
+      solver->setB(b.trilinos_rcp());
+      // First set whether we want to print the solver information to screen
+      // or not.
+      ConditionalOStream verbose_cout(std::cout, output_solver_details);
+
+      verbose_cout << "Starting solve" << std::endl;
+      solver->solve();
+
+      // Finally, force the SolverControl object to report convergence
+      solver_control.check(0, 0);
+      if (solver_control.last_check() != SolverControl::success)
+        AssertThrow(false,
+                    SolverControl::NoConvergence(solver_control.last_step(),
+                                                 solver_control.last_value()));
+    }
+
+    template <typename Number, typename MemorySpace>
+    void
+    SolverDirectBase<Number, MemorySpace>::do_solve()
+    {
+      // First set whether we want to print the solver information to screen or
+      // not.
+      ConditionalOStream verbose_cout(std::cout, output_solver_details);
+
+      solver->setParameters(Teuchos::rcpFromRef(parameter_list));
+      // Now do the actual factorization, which is a two step procedure.
+      // The symbolic factorization determines the structure of the inverse,
+      // while the numeric factorization does to actual computation of L and U
+      verbose_cout << "Starting symbolic factorization" << std::endl;
+      solver->symbolicFactorization();
+
+      verbose_cout << "Starting numeric factorization" << std::endl;
+      solver->numericFactorization();
+
+      verbose_cout << "Starting solve" << std::endl;
+      solver->solve();
+
+      // Finally, force the SolverControl object to report convergence
+      solver_control.check(0, 0);
+
+      if (solver_control.last_check() != SolverControl::success)
+        AssertThrow(false,
+                    SolverControl::NoConvergence(solver_control.last_step(),
+                                                 solver_control.last_value()));
+    }
+
+    template <typename Number, typename MemorySpace>
+    void
+    SolverDirectBase<Number, MemorySpace>::solve(
+      const SparseMatrix<Number, MemorySpace> &A,
+      Vector<Number, MemorySpace>             &x,
+      const Vector<Number, MemorySpace>       &b)
+    {
+      solver =
+        Amesos2::create<typename SparseMatrix<Number, MemorySpace>::MatrixType,
+                        MultiVector>(solver_type,
+                                     A.trilinos_rcp(),
+                                     x.trilinos_rcp(),
+                                     b.trilinos_rcp());
+      do_solve();
+    }
+
+    template <typename Number, typename MemorySpace>
+    SolverDirect<Number, MemorySpace>::AdditionalData::AdditionalData(
+      const std::string &solver_name,
+      const bool         output_solver_details)
+      : solver_name(solver_name)
+      , output_solver_details(output_solver_details)
+    {}
+
+    template <typename Number, typename MemorySpace>
```

Comment:
Our usual convention is 3 empty lines between functions. This whole file doesn't satisfy this.

Reply:
Done